### ISSUE TITLE
Handling HTML character references in embedded JSON-LD

### ISSUE THREAD

**Issue by gkellogg** [w3c/json-ld-syntax#100]
PR #97 and w3c/json-ld-api#51 resolved embedding JSON-LD in provide for HTML character references (entities) to be encoded within a data block so that critical elements are not mishandled by HTML parsers. As written this requires authors to encode sequences such as `<script></script>` which may occur in JSON-LD strings, and for these to be decoded when being extracted by the API.

It seems that the structured-data testing tool does not do this, and it's not clear what provisions are made for potentially problematic content.

One way to resolve this would be to provide direction to content authors to avoid any JSON-LD content embedded in HTML which could cause such a problem, which seems to be current practice. If character references are encoded, decoding is required at some layer.

cc/ @danbri @BigBlueHat

**Comment by danbri** [comment]
re Google SDTT, can you point me (well, danbri@google.com) at a testcase
and its expected vs actual outcome?

**Comment by gkellogg** [comment]
For example, consider the following JSON-LD:

```json
{
  "http://schema.org/description": "You can't embed a <script></script> snippet in JSON-LD embedded in HTML"
}
```

Embedding this in HTML creates errors due to illegal data-block content:

```html
<script type="application/ld+json">
{
  "http://schema.org/description": "You can't embed a <script></script> snippet in JSON-LD embedded in HTML"
}
</script>
```

As noted in [Restrictions for contents of JSON-LD `script` elements](https://w3c.github.io/json-ld-syntax/#restrictions-for-contents-of-json-ld-script-elements), such data must have HTML Character references encoded, such as the following:

```html
<script type="application/ld+json">
{
  "http://schema.org/description": "You can't embed a &lt;script&gt;&lt;/script&gt; snippet in JSON-LD embedded in HTML"
}
</script>
```

This feature is at risk, as it generates different results after encoding, unless the algorithm also decodes. It's apparent that the SDTT does not do such decoding.

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2018/2018-12-07-json-ld#section6-1).

- `RESOLVED:  close #100 with note about authoring concern related to HTML characters and </script> and remove existing text on JSON-LD processors handling character encodings`
<details><summary><i class="differentiate">View the transcript</i></summary>
6.1. Handling HTML character references in embedded JSON-LD<br/>
Benjamin Young: <a href="https://github.com/w3c/json-ld-syntax/issues/100">https://github.com/w3c/json-ld-syntax/issues/100</a><br/>
Benjamin Young: so far most of work has been Greg and Dan<br/>
Gregg Kellogg: this is a relatively narrow issue - we should solve by not resolving html entities in the API<br/>
… the narrow use case is script tags and comments<br/>
… so we should document by clarifying in spec language but not resolving the html entities<br/>
… if an html embedded in json (which is inside a script tag) closes the script tag, then rest of json will be treated as body<br/>
… in theory you can end up embedding javascript<br/>
… I suggest we document and say don’t do this<br/>
… and leave browser to defend against this<br/>
Benjamin Young: agree this seems not our problem<br/>
… people who put end script tag in their json-ld will find that they’ve broken their html<br/>
… if people need to escape character entities, they’ll do, but it will become an html issue not our problem<br/>
… while we don’t want to avoid edge cases, this still doesn’t seem to be our issue<br/>
… but would love to her back from Dan on this<br/>
… just add note that you could disrupt html parsing by doing this and leave it at that<br/>
Adam Soroka: I don’t want to minimize the security concern, but since this is coming from server side, it seems not as threatening<br/>
… not as bad from accepting user input<br/>
Gregg Kellogg: I would propose that we close this issue and remove all except a non-normative caution note<br/>
… that data block should remain valid<br/>
… I don’t anticipate anything different on this from schema.org<br/>
… this removes some text from API doc and maybe some changes in syntax document where we currently talk about how to escape<br/>
Proposed resolution: close #100 with note about authoring concern related to HTML characters and &lt;/script> and remove existingtext on JSON-LD processors handling character encodings (Benjamin Young)<br/>
Gregg Kellogg: +1<br/>
Tim Cole: +1<br/>
Benjamin Young: +1<br/>
Ivan Herman: +1<br/>
Adam Soroka: +1<br/>
David Newbury: +1<br/>
Pierre-Antoine Champin: +1<br/>
Resolution #3: close #100 with note about authoring concern related to HTML characters and </script> and remove existing text on JSON-LD processors handling character encodings<br/>
</details>

**Comment by gkellogg** [comment]
Reopening until edits are done.

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2018/2018-12-07-json-ld#section6-1).

- `RESOLVED:  close #100 with note about authoring concern related to HTML characters and </script> and remove existing text on JSON-LD processors handling character encodings`
<details><summary><i class="differentiate">View the transcript</i></summary>
Benjamin Young: so far most of work has been Greg and Dan<br/>
Gregg Kellogg: this is a relatively narrow issue - we should solve by not resolving html entities in the API<br/>
… the narrow use case is script tags and comments<br/>
… so we should document by clarifying in spec language but not resolving the html entities<br/>
… if an html embedded in json (which is inside a script tag) closes the script tag, then rest of json will be treated as body<br/>
… in theory you can end up embedding javascript<br/>
… I suggest we document and say don’t do this<br/>
… and leave browser to defend against this<br/>
Benjamin Young: agree this seems not our problem<br/>
… people who put end script tag in their json-ld will find that they’ve broken their html<br/>
… if people need to escape character entities, they’ll do, but it will become an html issue not our problem<br/>
… while we don’t want to avoid edge cases, this still doesn’t seem to be our issue<br/>
… but would love to her back from Dan on this<br/>
… just add note that you could disrupt html parsing by doing this and leave it at that<br/>
Adam Soroka: I don’t want to minimize the security concern, but since this is coming from server side, it seems not as threatening<br/>
… not as bad from accepting user input<br/>
Gregg Kellogg: I would propose that we close this issue and remove all except a non-normative caution note<br/>
… that data block should remain valid<br/>
… I don’t anticipate anything different on this from schema.org<br/>
… this removes some text from API doc and maybe some changes in syntax document where we currently talk about how to escape<br/>
Proposed resolution: close #100 with note about authoring concern related to HTML characters and &lt;/script> and remove existingtext on JSON-LD processors handling character encodings (Benjamin Young)<br/>
Gregg Kellogg: +1<br/>
Tim Cole: +1<br/>
Benjamin Young: +1<br/>
Ivan Herman: +1<br/>
Adam Soroka: +1<br/>
David Newbury: +1<br/>
Pierre-Antoine Champin: +1<br/>
Resolution #3: close #100 with note about authoring concern related to HTML characters and </script> and remove existing text on JSON-LD processors handling character encodings<br/>
</details>
